tests/robustness/pathlock/path_lock_test.go

```diff
@@ -0,0 +1,301 @@
+package pathlock
+
+import (
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestPathLockBasic(t *testing.T) {
+	pl := NewLocker()
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("Could not get working directory: %v", err)
+	}
+
+	for ti, tc := range []struct {
+		name  string
+		path1 string
+		path2 string
+	}{
+		{
+			name:  "(Abs) Blocks a Lock call for the same path /a/b/c",
+			path1: "/a/b/c",
+			path2: "/a/b/c",
+		},
+		{
+			name:  "(Abs) Blocks a Lock call for path /a/b/c/d",
+			path1: "/a/b/c",
+			path2: "/a/b/c/d",
+		},
+		{
+			name:  "(Abs) Blocks a Lock call for path /a/b",
+			path1: "/a/b/c",
+			path2: "/a/b",
+		},
+		{
+			name:  "(Abs) Blocks a Lock call for path /a",
+			path1: "/a/b/c",
+			path2: "/a",
+		},
+		{
+			name:  "(Rel) Blocks a Lock call for the same path a/b/c",
+			path1: "a/b/c",
+			path2: "a/b/c",
+		},
+		{
+			name:  "(Rel) Blocks a Lock call for path a/b/c/d",
+			path1: "a/b/c",
+			path2: "a/b/c/d",
+		},
+		{
+			name:  "(Rel) Blocks a Lock call for path a/b",
+			path1: "a/b/c",
+			path2: "a/b",
+		},
+		{
+			name:  "(Rel) Blocks a Lock call for path a",
+			path1: "a/b/c",
+			path2: "a",
+		},
+		{
+			name:  "(Mix Abs/Rel) Blocks a Lock call for the same path a/b/c",
+			path1: filepath.Join(cwd, "a/b/c"),
+			path2: "a/b/c",
+		},
+		{
+			name:  "(Mix Abs/Rel) Blocks a Lock call for path a/b/c/d",
+			path1: filepath.Join(cwd, "a/b/c"),
+			path2: "a/b/c/d",
+		},
+		{
+			name:  "(Mix Abs/Rel) Blocks a Lock call for path a/b",
+			path1: filepath.Join(cwd, "a/b/c"),
+			path2: "a/b",
+		},
+		{
+			name:  "(Mix Abs/Rel) Blocks a Lock call for path a",
+			path1: filepath.Join(cwd, "a/b/c"),
+			path2: "a",
+		},
+		{
+			name:  "(Mix Rel/Abs) Blocks a Lock call for the same path a/b/c",
+			path1: "a/b/c",
+			path2: filepath.Join(cwd, "a/b/c"),
+		},
+		{
+			name:  "(Mix Rel/Abs) Blocks a Lock call for path a/b/c/d",
+			path1: "a/b/c",
+			path2: filepath.Join(cwd, "a/b/c/d"),
+		},
+		{
+			name:  "(Mix Rel/Abs) Blocks a Lock call for path a/b",
+			path1: "a/b/c",
+			path2: filepath.Join(cwd, "a/b"),
+		},
+		{
+			name:  "(Mix Rel/Abs) Blocks a Lock call for path a",
+			path1: "a/b/c",
+			path2: filepath.Join(cwd, "a"),
+		},
+	} {
+		t.Logf("%v %v (path1: %q, path2: %q)", ti, tc.name, tc.path1, tc.path2)
+
+		lock1, err := pl.Lock(tc.path1)
+		if err != nil {
+			t.Fatalf("Unexpected path lock error: %v", err)
+		}
+
+		triggerCh := make(chan struct{})
+		trigger := false
+		var path2Err error
+
+		go func() {
+			lock2, err := pl.Lock(tc.path2)
+			if err != nil {
+				path2Err = err
+				close(triggerCh)
+				return
+			}
+
+			trigger = true
+			close(triggerCh)
+			lock2.Unlock()
+		}()
+
+		time.Sleep(10 * time.Millisecond)
+
+		if trigger == true {
+			t.Fatalf("Lock unsuccessful")
+		}
```

Comment:
There is no guarantee that the goroutine has even started execution after the sleep on line 130 has returned.  How about something like:
```
lockPathCh := make(chan struct{})
noLockPathCh := make(chan struct{})
go func() {
    l, err := pl.Lock(tc.path2)
    if err != nil {
        path2Err = err
        close(noLockPathCh)
        return
    }
    l.Unlock()
    close(lockPathCh)
}()
switch {
case <-noLockPathCh:
case <-lockPathCh:
      t.Fatalf("Lock unsuccessful")
}
```

Reply:
You could also use a single channel and return an `int` with a lock/nolock value.

Reply:
That should deadlock if the path lock is working correctly. Path1 has been locked and is expected to block the lock to Path2. Thus the `lockPathCh` will never close and we're stuck on the switch.

Reply:
Ah, yes! You're right.

Reply:
This calls for some "whitebox" testing:
```
syncCh :=. make(chan struct{})
go func() {
    close(syncCh)
    ... // as written
}()
<-syncCh // now we know the go routine has started
for {
    pl.mu.Lock()
    // some along these lines...
    lc, ok := pl.lockedPaths[path]
   pl.mu.Unlock()
    if ok {
       break // no error because the goroutine is blocked
   }
    if not timed out {
        sleep a short while
   } else { 
       error
   }
}
```

Reply:
The only way to really know that something is waiting for a lock would be to keep a counter in `pathLock` - e.g. an atomic that can be incremented before the call to `<- ch` and decremented after.
It would exist only for the UT but the UT is very important!

Reply:
Done. Added an atomic increment just before the channel wait. PTAL